File: py/src/imbrace/server/resources/categories.py
from typing import Any, Dict, List, Optional
from ...core.http import HttpTransport, AsyncHttpTransport


class CategoriesResource:
    """Categories domain — Server Gateway. /3rd/categories"""

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = f"{base}/3rd/categories"

    def list(self, organization_id: str) -> List[Dict[str, Any]]:
        return self._http.request("GET", self._base, params={"organization_id": organization_id}).json()

    def get(self, category_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._base}/{category_id}").json()

    def create(self, organization_id: str, name: str,
               apply_to: Optional[List[str]] = None,
               description: Optional[str] = None) -> Dict[str, Any]:
        body: Dict[str, Any] = {"name": name}
        if apply_to is not None:
            body["apply_to"] = apply_to
        if description is not None:
            body["description"] = description
        return self._http.request("POST", self._base, json=body,
                                  headers={"organization_id": organization_id}).json()

    def update(self, organization_id: str, category_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PUT", f"{self._base}/{category_id}", json=body,
                                  headers={"organization_id": organization_id}).json()

    def delete(self, category_id: str) -> Dict[str, Any]:
        return self._http.request("DELETE", f"{self._base}/{category_id}").json()


class AsyncCategoriesResource:
    """Categories domain — Server Gateway. Async."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = f"{base}/3rd/categories"

    async def list(self, organization_id: str) -> List[Dict[str, Any]]:
        res = await self._http.request("GET", self._base, params={"organization_id": organization_id})
        return res.json()

    async def get(self, category_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._base}/{category_id}")
        return res.json()

    async def create(self, organization_id: str, name: str,
                     apply_to: Optional[List[str]] = None,
                     description: Optional[str] = None) -> Dict[str, Any]:
        body: Dict[str, Any] = {"name": name}
        if apply_to is not None:
            body["apply_to"] = apply_to
        if description is not None:
            body["description"] = description
        res = await self._http.request("POST", self._base, json=body,
                                       headers={"organization_id": organization_id})
        return res.json()

    async def delete(self, category_id: str) -> Dict[str, Any]:
        res = await self._http.request("DELETE", f"{self._base}/{category_id}")
        return res.json()

File: ts/src/resources/categories.ts
import { HttpTransport } from "../http.js"

export interface Category {
  _id: string
  name: string
  apply_to?: string[]
  description?: string
  organization_id?: string
}

export interface CreateCategoryInput {
  name: string
  apply_to?: string[]
  description?: string
}

export interface UpdateCategoryInput {
  name?: string
  apply_to?: string[]
  description?: string
}

export class CategoriesResource {
  constructor(private readonly http: HttpTransport, private readonly base: string) {}

  private get url() { return `${this.base}/v1/backend/categories` }

  async list(organizationId: string): Promise<Category[]> {
    const url = new URL(this.url)
    url.searchParams.set("organization_id", organizationId)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async get(categoryId: string): Promise<Category> {
    return this.http.getFetch()(`${this.url}/${categoryId}`, { method: "GET" }).then(r => r.json())
  }

  async create(organizationId: string, body: CreateCategoryInput): Promise<Category> {
    return this.http.getFetch()(this.url, {
      method: "POST",
      headers: { "Content-Type": "application/json", "organization_id": organizationId },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async update(organizationId: string, categoryId: string, body: UpdateCategoryInput): Promise<Category> {
    return this.http.getFetch()(`${this.url}/${categoryId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json", "organization_id": organizationId },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async delete(categoryId: string): Promise<{ success: boolean }> {
    return this.http.getFetch()(`${this.url}/${categoryId}`, { method: "DELETE" }).then(r => r.json())
  }
}
